src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -2467,41 +2561,28 @@ public override BoundNode VisitMethodGroup(BoundMethodGroup node)
             return null;
         }
 
-        /// <summary>
-        /// Unconditionally visits an expression.
-        /// If the expression has "state when not null" after visiting,
-        /// the method returns 'true' and writes the state to <paramref name="stateWhenNotNull" />.
-        /// </summary>
-        protected bool VisitPossibleConditionalAccess(BoundExpression node, [NotNullWhen(true)] out TLocalState? stateWhenNotNull)
+        private bool TryVisitConditionalAccess(BoundExpression node, [NotNullWhen(true)] out TLocalState? stateWhenNotNull)
         {
-            EnterRegionIfNeeded(node);
-            var hasStateWhenNotNull = visit(out stateWhenNotNull);
-            Debug.Assert(!IsConditionalState);
-            LeaveRegionIfNeeded(node);
-            return hasStateWhenNotNull;
-
-            bool visit([NotNullWhen(true)] out TLocalState? stateWhenNotNull)
+            var access = node switch
             {
-                var access = node switch
-                {
-                    BoundConditionalAccess ca => ca,
-                    BoundConversion { Conversion: Conversion innerConversion, Operand: BoundConditionalAccess ca } when isAcceptableConversion(ca, innerConversion) => ca,
-                    _ => null
-                };
+                BoundConditionalAccess ca => ca,
+                BoundConversion { Conversion: Conversion innerConversion, Operand: BoundConditionalAccess ca } when isAcceptableConversion(ca, innerConversion) => ca,
+                _ => null
+            };
 
-                if (access is not null)
-                {
-                    return VisitConditionalAccess(access, out stateWhenNotNull);
-                }
-                else
-                {
-                    stateWhenNotNull = default;
-                    VisitWithStackGuard(node);
-                    Unsplit();
-                    return false;
-                }
+            if (access is not null)
+            {
+                EnterRegionIfNeeded(access);
+                Unsplit();
```

Comment:
Is this unsplit needed?

Reply:
Yes, this method can be called while in a conditional state. For example, `EqualsCondAccess_01` does this.

Consider an expression like `a && b(out x) == true`. We are in a conditional state after visiting the left. If the right is a conditional access, we have to unsplit that state here and the conditional state of the left is lost. But if the right is not a conditional access, we have to preserve the conditional state of the left in this method. This is because the caller may need to propagate that conditional state, like in the case where the right is a bool constant.

Reply:
I didn't follow. Let's sync offline

---
In reply to: [610078929](https://github.com/dotnet/roslyn/pull/52425#discussion_r610078929) [](ancestors = 610078929)